<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
   <head>
      <title>Results for 131217_PINKERTON_0286_AC3083ACXX Lane 2</title>
      <style type="text/css">
        * {font-family:Arial,Helvetica,sans-serif;}
      </style>
   </head>
   <body>
      <h2>Results for 131217_PINKERTON_0286_AC3083ACXX Lane 2</h2>
      <h3>Contents</h3>
      <ol>
      <li><a href="#stats">Sequencing Statistics</a></li>
      <li><a href="#data">Data Files</a></li>
      <li><a href="#mismatch">Mismatches vs Read Cycle</a></li>
      <li><a href="#quality">Quality Score vs Read Cycle</a></li>
      <li><a href="#calls">Base Call Composition vs Read Cycle</a></li>
      <li><a href="#intensity">Image Intensity vs Read Cycle</a></li>
      </ol>
      <a name="stats"><h3>Sequencing Statistics</h3></a>
      <p>
<p>
Run Information:
<table border="1" cellspacing="0" cellpadding="4">
<tr>
<td>Run Name</td>
<td>131217_PINKERTON_0286_AC3083ACXX</td>
</tr>
<tr>
<td>Flow Cell</td>
<td>C3083</td>
</tr>
<tr>
<td>Lane</td>
<td>2</td>
</tr>
<tr>
<td>Sample</td>
<td>GM19137_Rep2 rcvd 11/5/13</td>
</tr>
<tr>
<td>Submitter</td>
<td>Can Cenik</td>
</tr>
<tr>
<td>Platform</td>
<td>HiSeq 2000</td>
</tr>
<tr>
<td>Run Type</td>
<td>Single Read, 36bp</td>
</tr>
<tr>
<td>Genome</td>
<td>Human Male (hg19)</td>
</tr>
</table>
</p>
<p>
Sequencing Results:
<table border="1" cellspacing="0" cellpadding="4">
<tr>
<td>Total Reads</td>
<td align=right>141,186,892</td>
<td></td>
</tr>
<tr>
<td>Post-Filter Reads</td>
<td align=right>134,671,161</td>
<td>95.4% of Total Reads</td>
</tr>
<tr>
<td>Failed Reads</td>
<td align=right>6,515,731</td>
<td>4.6% of Total Reads</td>
</tr>
<tr>
<td>Mapped PF Reads</td>
<td align=right>37,589,590</td>
<td>27.9% of Post-Filter Reads</td>
</tr>
<tr>
<td>Uniquely-Mapped PF Reads</td>
<td align=right>17,536,443</td>
<td>13.0% of Post-Filter Reads</td>
</tr>
</table>
</p>
<p>
Detailed Alignment Results:
<table border="1" cellspacing="0" cellpadding="4">
<tr>
<td></td>
<td>Post-Filter Reads</td>
<td>% of Post-Filter Reads</td>
</tr>
<tr>
<td>Reads</td>
<td align=right>134,671,161</td>
<td></td>
</tr>
<tr>
<td>Mapped</td>
<td align=right>37,589,590</td>
<td align=right>27.9%</td>
</tr>
<tr>
<td>Unique</td>
<td align=right>17,536,443</td>
<td align=right>13.0%</td>
</tr>
<tr>
<td>Unique 0mm</td>
<td align=right>6,336,554</td>
<td align=right>4.7%</td>
</tr>
<tr>
<td>Unique 1mm</td>
<td align=right>6,407,309</td>
<td align=right>4.8%</td>
</tr>
<tr>
<td>Unique 2mm</td>
<td align=right>4,600,019</td>
<td align=right>3.4%</td>
</tr>
<tr>
<td>Unique 3+mm</td>
<td align=right>0</td>
<td align=right>0.0%</td>
</tr>
<tr>
<td>Unique w/Indel</td>
<td align=right>192,561</td>
<td align=right>0.1%</td>
</tr>
<tr>
<td>Non-Unique</td>
<td align=right>20,053,147</td>
<td align=right>14.9%</td>
</tr>
<tr>
<td>Non-Unique 0mm</td>
<td align=right>8,738,904</td>
<td align=right>6.5%</td>
</tr>
<tr>
<td>Non-Unique 1mm</td>
<td align=right>7,341,941</td>
<td align=right>5.5%</td>
</tr>
<tr>
<td>Non-Unique 2mm</td>
<td align=right>3,582,793</td>
<td align=right>2.7%</td>
</tr>
<tr>
<td>Non-Unique 3+mm</td>
<td align=right>0</td>
<td align=right>0.0%</td>
</tr>
<tr>
<td>Non-Unique w/Indel</td>
<td align=right>389,509</td>
<td align=right>0.3%</td>
</tr>
<tr>
<td>No-Match</td>
<td align=right>97,081,571</td>
<td align=right>72.1%</td>
</tr>
<tr>
<td>Contam</td>
<td align=right>N/A</td>
<td></td>
</tr>
<tr>
<td>Unclassified</td>
<td align=right>N/A</td>
<td></td>
</tr>
</table>
<p>
Key for the statistics tables:
<br />
Post-Filter = reads that passed the base-caller's quality filter
<br />
Failed = reads that failed the base-caller's quality filter
<br />
Unique = reads that mapped to a single position in the genome
<br />
Non-Unique = reads that mapped to more than one position
<br />
No-Match = reads that did not map anywhere
<br />
0mm = zero mismatches (perfect match to the genome)
<br />
1mm = one mismatching base
<br />
2mm = two mismatching bases
<br />
3+mm = three or more mismatching bases
<br />
w/Indel = alignment contains an insertion or deletion
<br />
Contam = reads that match the sequencing linkers
<br />
Unclassified = read was not mapped
</p>
      </p>
      <a name="data"><h3>Data Files</h3></a>
      <p>
<p>
Sequencing and analysis data may be downloaded using 
the links in this section.  The URLs are also included 
in the statistics CSV file below.  If you have an 
account on the SCG1 compute cluster then you can use 
the path names in the CSV file instead of the links.
</p>
<p>
<b><a href="131217_PINKERTON_0286_AC3083ACXX_L2_stats.csv">
Statistics and URLs (CSV File)</a></b>
</p>
<p>
Post-Filter Data
<table border="1" cellspacing="0" cellpadding="4">
<tr>
<td>Description</td>
<td>Barcode</td>
<td>Read Number</td>
<td>Format</td>
<td>File Size (Bytes)</td>
<td>Link</td>
</tr>
<tr>
<td>Unmapped Post-Filter Reads</td>
<td></td>
<td></td>
<td>FASTQ</td>
<td align=right>2,839,289,322</td>
<td><a href="http://scg-data.stanford.edu/Archive/IlluminaRuns/2013/dec/131217_PINKERTON_0286_AC3083ACXX/L2/131217_PINKERTON_0286_AC3083ACXX_L2_pf.fastq.gz">Download</a></td>
</tr>
<tr>
<td>Mapped Post-Filter Reads</td>
<td></td>
<td></td>
<td>BAM</td>
<td align=right>3,191,605,639</td>
<td><a href="http://scg-data.stanford.edu/Archive/IlluminaRuns/2013/dec/131217_PINKERTON_0286_AC3083ACXX/L2/131217_PINKERTON_0286_AC3083ACXX_L2_pf.bam">Download</a></td>
</tr>
<tr>
<td>Mapped Post-Filter Reads</td>
<td></td>
<td></td>
<td>BED</td>
<td align=right>162,640,519</td>
<td><a href="http://scg-data.stanford.edu/Archive/IlluminaRuns/2013/dec/131217_PINKERTON_0286_AC3083ACXX/L2/131217_PINKERTON_0286_AC3083ACXX_L2_pf.bed.gz">Download</a></td>
</tr>
<tr>
<td>Post-Filter Signal</td>
<td></td>
<td></td>
<td>SGR</td>
<td align=right>9,526,999</td>
<td><a href="http://scg-data.stanford.edu/Archive/IlluminaRuns/2013/dec/131217_PINKERTON_0286_AC3083ACXX/L2/131217_PINKERTON_0286_AC3083ACXX_L2_pf.sgr.gz">Download</a></td>
</tr>
</table>
</p>
<p>
Failed Data
<table border="1" cellspacing="0" cellpadding="4">
<tr>
<td>Description</td>
<td>Barcode</td>
<td>Read Number</td>
<td>Format</td>
<td>File Size (Bytes)</td>
<td>Link</td>
</tr>
<tr>
<td>Unmapped Failed Reads</td>
<td></td>
<td></td>
<td>FASTQ</td>
<td align=right>183,639,139</td>
<td><a href="http://scg-data.stanford.edu/Archive/IlluminaRuns/2013/dec/131217_PINKERTON_0286_AC3083ACXX/L2/131217_PINKERTON_0286_AC3083ACXX_L2_reject.fastq.gz">Download</a></td>
</tr>
</table>
</p>
      </p>
      <a name="mismatch"><h3>Mismatches vs Read Cycle</h3></a>
      <p>
<p>
This plot shows the fraction of bases that mismatch 
the reference sequence 
for each sequencing cycle (i.e. each position in the read). 
Vertical lines show boundaries between the reads if the run 
has multiple reads. 
An <a href="131217_PINKERTON_0286_AC3083ACXX_L2_details.html#mismatch">expanded plot</a> 
is also available (useful for longer read lengths).
</p>
        <img src="131217_PINKERTON_0286_AC3083ACXX_L2_mm_summary.png" alt="plot" />
      </p>
      <a name="quality"><h3>Quality Score vs Read Cycle</h3></a>
      <p>
<p>
The black points show the median quality score 
for each sequencing cycle (i.e. each position in the read). 
The gray lines show the 25% and 75% quartiles. 
Vertical lines show boundaries between the reads if the run 
has multiple reads. 
An <a href="131217_PINKERTON_0286_AC3083ACXX_L2_details.html#quality">expanded plot</a> 
is also available (useful for longer read lengths).
</p>
        <img src="131217_PINKERTON_0286_AC3083ACXX_L2_qual_summary.png" alt="plot" />
      </p>
      <a name="calls"><h3>Base Call Composition vs Read Cycle</h3></a>
      <p>
<p>
This plot shows the base composition for each sequencing 
cycle (i.e. each position in the read).  The vertical axis 
is the percentage for each base relative to all called bases. 
Vertical lines show boundaries between the reads if the run 
has multiple reads. 
An <a href="131217_PINKERTON_0286_AC3083ACXX_L2_details.html#calls">expanded plot</a> 
is also available (useful for longer read lengths).
</p>
        <img src="131217_PINKERTON_0286_AC3083ACXX_L2_call_summary.png" alt="plot" />
      </p>
      <a name="intensity"><h3>Image Intensity vs Read Cycle</h3></a>
      <p>
<p>
This plot shows the average image intensities 
for each sequencing cycle (i.e. each position in the read). 
These intensities have been corrected for cross talk 
and amplitude differences and are the input to the 
base caller. 
Vertical lines show boundaries between the reads if the run 
has multiple reads. 
An <a href="131217_PINKERTON_0286_AC3083ACXX_L2_details.html#intensity">expanded plot</a> 
is also available (useful for longer read lengths).
</p>
        <img src="131217_PINKERTON_0286_AC3083ACXX_L2_int_summary.png" alt="plot" />
      </p>
<p>
Additional <a href="131217_PINKERTON_0286_AC3083ACXX_L2_diags.html">Diagnostic Plots</a>
</p>
   </body>
</html>
